docs/proposals/component-registration.md

```diff
@@ -0,0 +1,409 @@
+
+As a service that provides reports/graphs (e.g. DataDog), I want to retrieve status metrics for each component, so that I can record and report on the metrics over time.
+
+### Self-Healing
+
+As a Kubernetes Operator, I want the cluster to self-heal when individual components crash so that I don't get notified or need to intervene.
+
+As a process monitoring tool (e.g. monit), I want to be able to determine the condition of the component process I am monitoring to determine if it needs to be restarted.
+
+As a container/vm monitoring tool (e.g. Marathon, Amazon CloudWatch), I want to be able to determine the health of the component in the container/vm I am monitoring.
+
+Note: Self-healing external components requires knowing how to deploy the external components, which is different for each cloud platform and changes based on component configuration. This is different than self-healing addon components, which is covered by the Kubernetes replication controller.
+
+
+## Design
+
+Unfortunately, ComponentStatuses is almost not worth the effort to fix. Instead it can be deprecated and replaced by a combination of Service and Endpoints changes. This allows the components and their addresses & ports to be specified more flexibly.
+
+Since it's seemingly impossible to fix Endpoints Problem \#1 without breaking reverse compatibility, this proposal punts on \#1 (until v2 allows reverse-incompatibility) and focuses on fixing \#2 for now. #13778 handles increasing the scope of Endpoints to cover both Ready and NotReady addresses, which is better than nothing, but not as flexible as having an array of conditions, like other -Status API resources.
+
+To fix Endpoints Problem \#2, this proposal moves the specification (and status) of external service provider addresses to its own namespaced API resource (provisionally called **ExternalServiceProvider**). Endpoints Controller will then continue to manage the lifecycle (create/update/delete) of Internal Endpoints based on a selection of Pod instances, but will also do the same for External Endpoints based on a selection of ExternalServiceProviders defined in ServiceSpec.Selector. Endpoints then becomes a (conceptually) read-only API for the user. Endpoints can't actually be read-only, because the Endpoints Controller needs access, but write access may eventually be limited to admin/operator-only.
+
+
+### Phased Changes
+
+Since the changes here are complex, they will be implemented in phases.
+
+1. Simplify Endpoints
+  - Add a new API resource (ExternalServiceProvider) that represents a non-pod service provider
+    - Specify a network route (host[:port])
+    - Specify liveness and readiness probe methods
+    - Store the last known status (list of conditions)
+  - Add (or modify) a controller to probe ExternalServiceProviders and update their status
+  - Update how Endpoints are managed
+    - EITHER update the EndpointsController to manage Endpoints based on both Pod and ExternalServiceProvider status
```

Comment:
Do you imagine needing both pods and imported endpoints in the same service? That seems like it could be confusing, though I'm not keen to add another API field to specify which one is expected.